### ISSUE TITLE
(A-)synchronization

### ISSUE THREAD

**Issue by philip98** [iApocalypsee/Pfeile#14]
I think, we should agree which _aspects_ of Pfeile should be handled synchronously and which should be handled asynchronously.

As far as I see, all input is handled by AWT's listener system and there is no way to circumvent this, so it is fine. However, all output, i.e. rendering, is done synchronously by `GameLoop` calling the current `GameWindow`'s `draw()` method, which then continues calling other `draw()` methods. I don't understand why we introduce this new hierarchy of nested `draw()` calls when there already exists a mechanism for the designed purpose of drawing/painting/whatever components and their children, namely the `paint()` mechanism. Further information on this subject can be found [here](http://www.oracle.com/technetwork/java/painting-140037.html).
The third _aspect_ that should be considered is state. @iApocalypsee already showed how to handle game state transitions synchronously and this should be done for all kinds of state by introducing a chain of `updateState()` methods (that take time as parameter).

**Comment by iApocalypsee** [comment]
I think we may never come to a conclusive agreement on what could and should be handled asynchronously; however, there are a few things you already talked about that definitely will not be executed in parallel, namely the game loop in general or the translation system. Everything else which is not part of the actual game loop is currently in shady parallel/non-parallel territory. Initialization of the game for example is done by a number of threads which ultimately rely on the other init threads to not interfere with their business, while the main thread simultaneously does some initialization of its own; this has to be addressed at some other point in time.

As for the separate `draw()` call hierarchy, I can only make assumptions on why it exists. The most important aspect has been to have ultimate _control_ over the draw calls and to not redirect this responsibility to AWT and its internals (which we did know very well back then and still do not).

I like the idea of update chains, and there were already some efforts made to establish that sort of thing (`interface GUIUpdateable` to name one failed attempt, which lacks the time parameter). We somehow ended up not using it, and I still do not know why.

**Comment by Request01** [comment]
The initialization process is completely threaded until LoadingWorldScreen/ContextCreator. The threads do NOT rely upon previous threads to finish anymore (except the possibility of entering LoadingWorldScreen [I don't want to search the code which screen right now] befor it is loaded and this is practically impossible, it just takes so long for the user to leave  both the PreWindowScreen and ArrowSelectionScreenPreSet.). If you know any issues with the initialization, please mention them, but as far as I know I solved any problems regarding this matter. 
I still belive

**Comment by iApocalypsee** [comment]
@Request01 This issue is not about the problems the initialization process currently has (for the record, it has close to none); it is rather about the uncertainty of threading or non-threading for some tasks performed by the game.

A proposal for a 'guideline' regarding synchronization of certain tasks:
 * Initialization: as much threading as possible, as long as maintainability is guaranteed.
 * Less threading while game loop is executed, should be handled mostly by the update loop. Only when it is necessary should be threads allowed to run then.
 * Cleanup may be executed asynchronously as well, although I currently have no clue on what to clean up right now.

**Comment by Request01** [comment]
The problem itself is that the game loop is already executed not fast enough. During difficult frames (e. g. when the world is moved), the decreased calculating speed of the CPU (compared to the GPU) is already visible. That's why we definitive have to thread what is possible and that encompasses the game loop as well, even though a single threaded handleing is safer and more convenient.

**Comment by iApocalypsee** [comment]
Ditch convenience? That I can do easily.
Good point actually; we are not computing the graphics on the graphics card but rather with processor cycles.
So as to speak, the general guideline would be to make everything run in threads more or less.
